File: backend/app/api/v1/schemas/staking.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class BuildStakingTransactionRequest(BaseModel):
    wallet_address: str = Field(min_length=10, max_length=128)
    action: Literal["delegate", "undelegate", "redelegate"]
    subnet_id: int = Field(ge=0)
    amount_rao: int = Field(gt=0)
    source_validator_hotkey: str | None = Field(default=None, max_length=128)
    dest_validator_hotkey: str | None = Field(default=None, max_length=128)


class StakingSimulationResponse(BaseModel):
    supported: bool
    status: str
    message: str
    estimated_fee_rao: int | None = None
    block_number: int | None = None


class StakingTransactionPreview(BaseModel):
    action: str
    subnet_id: int
    amount_rao: int
    source_validator_hotkey: str | None
    dest_validator_hotkey: str | None
    estimated_fee_rao: int | None = None


class StakingTransactionResponse(BaseModel):
    id: UUID
    wallet_address: str
    action: str
    subnet_id: int
    amount_rao: int
    source_validator_hotkey: str | None
    dest_validator_hotkey: str | None
    status: str
    unsigned_payload: dict[str, object]
    simulation: StakingSimulationResponse | None
    tx_hash: str | None
    failure_reason: str | None
    expires_at: datetime
    submitted_at: datetime | None
    confirmed_at: datetime | None
    preview: StakingTransactionPreview


class SubmitStakingTransactionRequest(BaseModel):
    tx_hash: str = Field(min_length=10, max_length=128)
    signed_extrinsic: str | None = None

File: frontend/lib/api/staking.ts
export type StakingAction = "delegate" | "undelegate" | "redelegate";

export type StakingSimulation = {
  supported: boolean;
  status: string;
  message: string;
  estimated_fee_rao?: number | null;
  block_number?: number | null;
};

export type StakingTransactionPreview = {
  action: StakingAction;
  subnet_id: number;
  amount_rao: number;
  source_validator_hotkey: string | null;
  dest_validator_hotkey: string | null;
  estimated_fee_rao?: number | null;
};

export type StakingTransaction = {
  id: string;
  wallet_address: string;
  action: StakingAction;
  subnet_id: number;
  amount_rao: number;
  source_validator_hotkey: string | null;
  dest_validator_hotkey: string | null;
  status: string;
  unsigned_payload: Record<string, unknown>;
  simulation: StakingSimulation | null;
  tx_hash: string | null;
  failure_reason: string | null;
  expires_at: string;
  submitted_at: string | null;
  confirmed_at: string | null;
  preview: StakingTransactionPreview;
};

export type BuildStakingTransactionInput = {
  wallet_address: string;
  action: StakingAction;
  subnet_id: number;
  amount_rao: number;
  source_validator_hotkey?: string;
  dest_validator_hotkey?: string;
};

export type PaginatedStakingTransactions = {
  data: StakingTransaction[];
  pagination: {
    page: number;
    page_size: number;
    total_items: number;
    total_pages: number;
  };
};
